documentation/Action-plugin-specification.md

```diff
@@ -0,0 +1,133 @@
+  - Templates store the actions associated with them by referencing this `code`
+- `name` -- the name displayed in the UI (mainly Template Builder)
+- `description` -- short explanation of the plugin's functionality. Also displayed in the UI.
+- `file` -- the name of the source code file. Doesn't include the file extension, as the dynamic import will read from either the `.js` or `.ts` file depending on context (development vs. build).
+- `function_name` -- the name of the function exported in the source code file. If this name is changed in the source file, it _must_ be updated here, or else the app won't find the function.
+- `required_parameters` -- an array of strings that define the names of the fields that must be supplied to the plugin at runtime. Templates store the values (or dynamic expressions) that map to these field names.
+
+## How plugins are loaded.
+
+At startup, the server app scans the `plugins` folder. In each plugin's subfolder, it looks for a `plugin.json` file, from which it reads the metadata. It compares this information with the list of plugins registered in `action_plugin` and updates the database accordingly:
+
+- Adds a new record for newly-found plugins
+- Updates plugin details that have changed
+- Removes records for missing plugins
+
+The `action_plugin` table is the primary record of what actions are available to the system, with the following fields stored:
+
+- `code`
+- `name`
+- `description`
+- `path` (determined by the folder the plugin was found in, plus the specified filename)
+- `function_name`
+- `required_parameters`
+
+<a name="development"></a>
+
+## Development and build process
+
+Plugins are intended to be stand-alone units, in that a user should be able to simply copy a plugin folder to another system and have it work straight away. (Eventually, we envisage that plugins could be imported directly via the front-end UI.) To that end, each plugin is developed as its own package, with its own `package.json` and `node_modules` folder, as well as Typescript and testing configurations.
+
+When developing a plugin, you can run the main app with `yarn dev` as normal, and changes you make to the plugin `.ts` file will be reflected immediately in the dev environment.
+
+However, when development work is complete, the plugin should be built independently by running `yarn build` _in the plugin's root folder_, which will compile the typescript code into `.js` file(s). (There is also a shortcut `yarn build_plugins` to build all plugins from the project root.) When the main project is built (`yarn build`), the plugins are not re-compiled, but simply copied directly to the build folder, so its expected they will each have their own compiled `.js` file already in place.
+
+## Needs consideration
+
+- Should the `required_parameters` specify an expected type so that the Template Builder UI knows what kind of options to present?
```

Comment:
Yes. I think so, we should give errors from what is selected to be queried for each field checking if the final result is the same as the expected.

Reply:
I think the main purpose of the "type" here would be so the UI can know what kind of options to make available. So, for example, if the required parameter was `email_address`, the expected type might be "email address", so then the UI in the Template Builder might present these options:
- user's email address (saves query to lookup `user.email`)
- other email (user fills in a free-text string, which is saved as a literal value.

And then there would be mappings stored somewhere that link "types" to "presentation options", or something like that. 

We should probably have a discussion around this stuff soon to get some clarification before getting in too deep.

Reply:
Yes. It would be better to add as part of next meeting. : )